File: Curso-ciencias-de-la-computacion/12.-Sorting/Quicksort/main.py
#Quicksort implementation in python
#By JCPC 12/11/2025
#time complexity average and best case: O(nlogn) worst case: O(n^2)

def quicksort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)

        quicksort(arr, low, pi-1)
        quicksort(arr, pi+1, high)

def partition(arr, low, high):
    pivot = arr[high]

    i = low-1

    #ordering values < pivot to the left
    for j in range(low, high):
        if arr[j] < pivot:
            i+=1
            arr[i], arr[j] = arr[j], arr[i]

    #putting pivot in its position, next to last low value
    arr[high], arr[i+1] = arr[i+1], arr[high]

    return i+1


def main():
    arr = [10, 7, 8, 9, 1, 5]
    n = len(arr)
    quicksort(arr, 0, n-1)
    print(arr)

if __name__=='__main__':
    main()

File: Curso-ciencias-de-la-computacion/12.-Sorting/Quicksort/CPP/main.cpp
#include <iostream>
#include <vector>

int partition(std::vector<int>&arr, int low, int high){
    int pivot = arr[high];

    int i = low-1;

    for(int j = low; j < high; j++){
        if(arr[j] < pivot){
            i++;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[high], arr[i+1]);
    return i + 1;
}

void quickSort(std::vector<int>& arr, int low, int high){
    if (high > low){

        int pi = partition(arr, low, high);

        quickSort(arr, low, pi-1);
        quickSort(arr, pi+1, high);
    }
}

int main(){
    std::vector<int> arr = {7,3,9,5,4,8,0,1};
    int n = arr.size();
    quickSort(arr, 0, n-1);
    for(int i = 0; i < n; i++){
        std::cout << arr[i] << " ";
    }
    return 0;
}